File: java_practice/InvertBinaryTree.java
package java_practice;

import java.util.*;

public class InvertBinaryTree {
     public TreeNode invertTree(TreeNode root) {
        if(root == null) return null;

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i = 0; i < size; i++){
                TreeNode node = queue.poll();
                // swap
                TreeNode tmpNode = node.left;
                node.left = node.right;
                node.right = tmpNode;
                // add left and right
                if(node.left!= null)queue.offer(node.left);
                if(node.right!=null)queue.offer(node.right);
            }
        }
        return root;
    }

    private void print(TreeNode root){
        if(root == null) System.out.println("[]");

        List<Integer> list = new ArrayList<>();
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i = 0; i < size; i++){
                TreeNode node = queue.poll();
                list.add(node.val);
                // add left and right
                if(node.left!= null)queue.offer(node.left);
                if(node.right!=null)queue.offer(node.right);
            }
        }
        System.out.println(list.toString());
    }

    public static void main(String[] args) {
        InvertBinaryTree solver = new InvertBinaryTree();
        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(2);
        root.right = new TreeNode(7);
        root.left.left = new TreeNode(1);
        root.left.right = new TreeNode(3);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(9);

        solver.print(root);
        solver.invertTree(root);
        solver.print(root);

    }
}

File: js_practice/tree/invert_binary_tree.js
class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

const invert_tree = (root) => {
  if (!root) return null;

  const tmp = root.left;

  root.left = root.right;
  root.right = tmp;

  if (root.left) invert_tree(root.left);
  if (root.right) invert_tree(root.right);

  return root;
};

const root = new TreeNode(
  4,
  new TreeNode(2, new TreeNode(1), new TreeNode(3)),
  new TreeNode(7, new TreeNode(6), new TreeNode(9)),
);

const printTree = (root) => {
  if (!root) return [];
  const res = [];
  const queue = [root];
  while (queue.length) {
    const node = queue.shift();
    if (node.left) queue.push(node.left);
    if (node.right) queue.push(node.right);
    res.push(node.val);
  }
  console.log(123123, res);
};

printTree(invert_tree(root));
